### PR TITLE
Django2 and Docker

### PR DESCRIPTION
**Pull request by miguelcleon**
@lsetiawan I made two minor changes to the dockerfile and I'm trying to run the build on dockerhub but I'm not really sure what I'm doing. https://hub.docker.com/r/odm2/odm2-admin/builds/

### CODE REVIEW

**Comment by lsetiawan** [comment]
@miguelcleon I just saw this. My apologies. I won't be able to work on this today, but I'll take a look tomorrow what's going on. Thanks. Does it work locally when you try to build the Docker image?

**Comment by miguelcleon** [comment]
@lsetiawan ok, sounds good thanks. No I haven't gotten to work. I'm trying to run it on a VM and I ran out of space on it and it is taking a long time to expand the drive. I changed the host name sent to postgresql in the latest commit to this PR. that seems to be the latest error.

**Comment by miguelcleon** [comment]
Now I get a different error in the docker log:

```
psql: could not connect to server: Connection refused
        Is the server running on host "127.0.0.1" and accepting
        TCP/IP connections on port 5432?
Postgres is unavailable - sleeping

```

**Comment by lsetiawan** [comment]
After looking at the code, I remember what I did. I separated the database from the actual application and use docker-compose to spin things up. See https://github.com/ODM2/ODM2-Admin/tree/master/docker/docker-deploy

This auto build the image. I guess the application docker file needs to be tweaked to be a stand alone rather than trying to attach to the non-existent database, which causes the error.

**Comment by miguelcleon** [comment]
@lsetiawan so how was it attaching to the database before? I'm not understanding what broke. Thanks

**Comment by lsetiawan** [comment]
Before it installs and spins up postgresql in the background.

**Comment by lsetiawan** [comment]
Have you tried doing `docker-compose up`?

File: docker/odm2admin/Dockerfile [hunk 1]
```diff
@@ -36,4 +36,4 @@ COPY startup.sh /startup.sh
 RUN chmod 755 /wait-for-postgres.sh
 RUN chmod 755 /startup.sh
 
-CMD ["/bin/bash", "/wait-for-postgres.sh", "db", "/startup.sh"]
+CMD ["/bin/bash", "/wait-for-postgres.sh", "127.0.0.1", "/startup.sh"]
```
Comment (lsetiawan) [inline docker/odm2admin/Dockerfile]: With `docker-compose up` you'd need to change this back to 'db'

Comment (miguelcleon) [reply docker/odm2admin/Dockerfile]: Ok, I'll try again, I was having trouble with docker in my local environment, trying it on a server now.

Comment (miguelcleon) [reply docker/odm2admin/Dockerfile]: OK, seems like I had installed the wrong version of docker so it took a little while to work around that. Docker-compose hung trying to install hs_restclient. I'll try again. 

```
oauthlib-2.0.6 100% |###############################| Time: 0:00:00  44.69 MB/s
urllib3-1.22-p 100% |###############################| Time: 0:00:00  56.83 MB/s
requests-2.18. 100% |###############################| Time: 0:00:00  59.81 MB/s
requests-oauth 100% |###############################| Time: 0:00:00 132.88 kB/s
requests-toolb 100% |###############################| Time: 0:00:00  45.24 MB/s
hs_restclient-  91% |#^CGracefully stopping... (press Ctrl+C again to force)B/s
```

Comment (miguelcleon) [reply docker/odm2admin/Dockerfile]: yeah, looks like the same thing is happening, I guess I'll comment out hs_restclient for now.

Comment (miguelcleon) [reply docker/odm2admin/Dockerfile]: @lsetiawan I'd have to post a change to requirements.txt into a release right? I guess I could do that on miguelcleon/ODM2-Admin

Comment (miguelcleon) [reply docker/odm2admin/Dockerfile]: probably better to figure out why hs_restclient won't download

Comment (lsetiawan) [reply docker/odm2admin/Dockerfile]: > @lsetiawan I'd have to post a change to requirements.txt into a release right? I guess I could do that on miguelcleon/ODM2-Admin

Yea you'd have too.

> probably better to figure out why hs_restclient won't download

Hmm.. that's weird.

Comment (miguelcleon) [reply docker/odm2admin/Dockerfile]: OK, the problem is I ran out of space on that drive. now I'm getting this error, which seems like progress: 

```
odm2adminapp | 
odm2adminapp | ERROR:  database "odm2" already exists
odm2adminapp |                                     List of databases
odm2adminapp |        Name       |  Owner   | Encoding |  Collate   |   Ctype    |   Access privileges   
odm2adminapp | ------------------+----------+----------+------------+------------+-----------------------
odm2adminapp |  odm2             | postgres | UTF8     | en_US.utf8 | en_US.utf8 | 
odm2adminapp |  postgres         | postgres | UTF8     | en_US.utf8 | en_US.utf8 | 
odm2adminapp |  template0        | postgres | UTF8     | en_US.utf8 | en_US.utf8 | =c/postgres          +
odm2adminapp |                   |          |          |            |            | postgres=CTc/postgres
odm2adminapp |  template1        | postgres | UTF8     | en_US.utf8 | en_US.utf8 | =c/postgres          +
odm2adminapp |                   |          |          |            |            | postgres=CTc/postgres
odm2adminapp |  template_postgis | postgres | UTF8     | en_US.utf8 | en_US.utf8 | 
odm2adminapp | (5 rows)
odm2adminapp | 
odm2adminapp | ERROR:  database "odm2" already exists
odm2admindb | ERROR:  database "odm2" already exists
odm2admindb | STATEMENT:  CREATE DATABASE odm2;
odm2adminapp exited with code 1
```

Comment (lsetiawan) [reply docker/odm2admin/Dockerfile]: Yepp that's a database problem. Glad you're making progress. 😄

Comment (miguelcleon) [reply docker/odm2admin/Dockerfile]: Yeah, I've got to move docker to a different drive. backing up now, which is taking a while. Docker is a disk drive hog.

Comment (lsetiawan) [reply docker/odm2admin/Dockerfile]: > Docker is a disk drive hog.

Yea you gotta clean up all the containers that failed when trying to build and images that dangles to get more space.

Comment (miguelcleon) [reply docker/odm2admin/Dockerfile]: Here is the problem or a problem. It looks like it is not pulling the latest release. Line 5 in development.py does not match https://github.com/ODM2/ODM2-Admin/blob/master/templatesAndSettings/settings/development.py#L5

```

odm2adminapp | Migrate django tables
odm2adminapp | Traceback (most recent call last):
odm2adminapp |   File "manage.py", line 10, in <module>
odm2adminapp |     execute_from_command_line(sys.argv)
odm2adminapp |   File "/opt/conda/envs/odm2adminenv/lib/python3.6/site-packages/django/core/management/__init__.py", line 371, in execute_from_command_line
odm2adminapp |     utility.execute()
odm2adminapp |   File "/opt/conda/envs/odm2adminenv/lib/python3.6/site-packages/django/core/management/__init__.py", line 317, in execute
odm2adminapp |     settings.INSTALLED_APPS
odm2adminapp |   File "/opt/conda/envs/odm2adminenv/lib/python3.6/site-packages/django/conf/__init__.py", line 56, in __getattr__
odm2adminapp |     self._setup(name)
odm2adminapp |   File "/opt/conda/envs/odm2adminenv/lib/python3.6/site-packages/django/conf/__init__.py", line 43, in _setup
odm2adminapp |     self._wrapped = Settings(settings_module)
odm2adminapp |   File "/opt/conda/envs/odm2adminenv/lib/python3.6/site-packages/django/conf/__init__.py", line 106, in __init__
odm2adminapp |     mod = importlib.import_module(self.SETTINGS_MODULE)
odm2adminapp |   File "/opt/conda/envs/odm2adminenv/lib/python3.6/importlib/__init__.py", line 126, in import_module
odm2adminapp |     return _bootstrap._gcd_import(name[level:], package, level)
odm2adminapp |   File "<frozen importlib._bootstrap>", line 994, in _gcd_import
odm2adminapp |   File "<frozen importlib._bootstrap>", line 971, in _find_and_load
odm2adminapp |   File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
odm2adminapp |   File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
odm2adminapp |   File "<frozen importlib._bootstrap_external>", line 678, in exec_module
odm2adminapp |   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
odm2adminapp |   File "/ODM2-Admin/templatesAndSettings/settings/development.py", line 5, in <module>
odm2adminapp |     from base import *
```

**Comment by miguelcleon** [comment]
@lsetiawan oh these settings files need to be updated https://github.com/ODM2/ODM2-Admin/blob/master/docker/docker-deploy/settings/

**Comment by lsetiawan** [comment]
> @lsetiawan oh these settings files need to be updated https://github.com/ODM2/ODM2-Admin/blob/master/docker/docker-deploy/settings/

Yes indeed.

**Comment by miguelcleon** [comment]
forgot to change the name of this last commit

**Comment by miguelcleon** [comment]
It seems like docker-compose should work but I keep getting: 

```
odm2admindb | ERROR:  database "odm2" already exists
odm2admindb | STATEMENT:  CREATE DATABASE odm2;
```
maybe if I comment out https://github.com/ODM2/ODM2-Admin/blob/master/docker/odm2admin/wait-for-postgres.sh#L15

**Comment by lsetiawan** [comment]
> maybe if I comment out https://github.com/ODM2/ODM2-Admin/blob/master/docker/odm2admin/wait-for-postgres.sh#L15

That should work, but it's weird how "odm2" database already exist. hmm...

**Comment by miguelcleon** [comment]
nope that didn't help

**Comment by miguelcleon** [comment]
I needed to change the git clone branch in the dockerfile but now I get:

odm2admindb | FATAL:  database "odm2" does not exist
odm2adminapp | psql: FATAL:  database "odm2" does not exist
odm2admindb | FATAL:  database "odm2" does not exist

so that  `CREATE DATABASE odm2;` needs to run the first time only in `wait-for-postgres.sh `

**Comment by miguelcleon** [comment]
maybe it should not stop on error, like this I guess: 

`psql -v ON_ERROR_STOP=0 --host "$host" --username "postgres" -c "CREATE DATABASE odm2;"`

**Comment by lsetiawan** [comment]
@miguelcleon What's the status of this?

**Comment by miguelcleon** [comment]
@lsetiawan seems like it works if I comment out: 
`psql -v ON_ERROR_STOP=1 --host "$host" --username "postgres" -c "CREATE DATABASE odm2;"`
from here:
https://github.com/ODM2/ODM2-Admin/blob/master/docker/odm2admin/wait-for-postgres.sh#L15

but only in the local copy; if I comment it out in github then I get a no odm2 error:
```

odm2admindb | FATAL: database "odm2" does not exist
odm2adminapp | psql: FATAL: database "odm2" does not exist
odm2admindb | FATAL: database "odm2" does not exist
```

So [this](https://github.com/ODM2/ODM2-Admin/blob/master/docker/odm2admin/wait-for-postgres.sh#L15) should only run the first time; maybe move it to the dockerfile?

**Comment by miguelcleon** [comment]
I think it works now. Building dockerhub repo based on miguelcleon fork https://hub.docker.com/r/miguelcleon/odm2-admin/builds/

**Comment by miguelcleon** [comment]
success! @lsetiawan maybe you could try running the docker container? https://hub.docker.com/r/miguelcleon/odm2-admin/ pinging @emiliom

**Comment by miguelcleon** [comment]
@lsetiawan did you get a chance to try this?

**Comment by lsetiawan** [comment]
Am I still composing up or simply run the docker image?

**Comment by miguelcleon** [comment]
Try both. I had different issues in actually running the app in a web browser that I'm confused about.

**Comment by lsetiawan** [comment]
I am having sql problem when creating the database using `docker-compose`:
```
odm2adminapp | psql:/odm2adminDB.sql:30: ERROR:  schema "admin" already exists
odm2admindb | ERROR:  schema "admin" already exists
odm2admindb | STATEMENT:  CREATE SCHEMA admin;
```

and getting 

```
psql: could not translate host name "db" to address: Name or service not known
Postgres is unavailable - sleeping
psql: could not translate host name "db" to address: Name or service not known
Postgres is unavailable - sleeping
```

because the application doesn't map to db. I think might have to put a Database option to the docker image.

**Comment by miguelcleon** [comment]
Are you using the django2 branch?

**Comment by lsetiawan** [comment]
Yepp. And used your docker image.

**Comment by miguelcleon** [comment]
Ok I guess we might have to put a Database option to the docker image

**Comment by miguelcleon** [comment]
@lsetiawan so how do we add the database option to the docker image? I've played around with this but have not had success.

**Comment by lsetiawan** [comment]
@miguelcleon I suggest using [Docker ENV](https://docs.docker.com/engine/reference/builder/#env) This sets the system environment variables for docker. So when running the images, you can do `docker run -e DBHOST=123.23.123.12 -e DBUSER=someone -e DBPASSWORD=blahblah ...`

Then id the development.py settings python file, should be able to read from `os.environ` with those credentials and host.

**Pull request merged**